File: client/src/types/commerce.ts
export type Money = { amount: number; currency: string };

export type ProductVariant = {
  id: number;
  title: string;
  sku: string;
  color: string;
  colorTr?: string | null;
  swatchHex?: string | null;
  price: number;
  priceOverride?: number | null;
  stock: number;
  lowStockThreshold: number;
  isActive: boolean;
};

export type ProductMedia = {
  id: number;
  imagePath: string;
  alt: string;
  sortOrder: number;
  isPrimary: boolean;
};

export type ProductModelAsset = {
  id: number;
  variantId?: number | null;
  glbPath: string;
  usdzPath?: string | null;
  posterPath?: string | null;
  alt: string;
  placement: 'floor' | 'wall';
  supportedPlacements: Array<'floor' | 'wall'>;
  scaleMode: 'fixed';
  dimensionsMm: { width: number; height: number; depth: number };
};

/**
 * Mirrors the API's ProductDto. Turkish copy travels on the record itself, so a product
 * created in the admin panel is localizable without touching the frontend.
 */
export type Product = {
  id: number;
  name: string;
  nameTr?: string | null;
  slug: string;
  category: string;
  categoryTr?: string | null;
  categorySlug: string;
  categoryId: number;
  description: string;
  descriptionTr?: string | null;
  story: string;
  storyTr?: string | null;
  material: string;
  materialTr?: string | null;
  dimensions: string;
  dimensionsTr?: string | null;
  price: Money;
  compareAt?: Money | null;
  image: string;
  images: string[];
  colors: string[];
  rating: number;
  reviews: number;
  stock: number;
  isNew: boolean;
  isFeatured: boolean;
  isActive: boolean;
  sku: string;
  seoTitle: string;
  metaDescription: string;
  variants: ProductVariant[];
  mediaItems: ProductMedia[];
  modelAssets: ProductModelAsset[];
  createdAt: string;
  updatedAt: string;
};

export type Category = {
  id: number;
  name: string;
  nameTr?: string | null;
  slug: string;
  description: string;
  descriptionTr?: string | null;
  image?: string | null;
  sortOrder: number;
  count: number;
};

export type CartItem = {
  id: number;
  variantId: number;
  productId: number;
  productName: string;
  productNameTr?: string | null;
  productSlug: string;
  sku: string;
  color: string;
  colorTr?: string | null;
  image?: string | null;
  quantity: number;
  unitPrice: number;
  lineTotal: number;
  availableStock: number;
  giftBoxKey?: string | null;
  giftMessage?: string | null;
  packagingNotes?: string | null;
  surpriseRecipient?: string | null;
  surpriseVibes?: string[];
  surpriseInstructions?: string | null;
};

export type Cart = {
  id: number;
  currency: string;
  items: CartItem[];
  subtotal: number;
  discountTotal: number;
  couponDiscountTotal: number;
  tradeInCredit: number;
  totalAfterDiscount: number;
  coupon?: AppliedCoupon | null;
  tradeIn?: TradeIn | null;
  count: number;
};

export type TradeInCondition = 'like_new' | 'good' | 'fair';
export type TradeInHandoffMethod = 'pickup' | 'drop_off';
export type TradeInStatus = 'estimate' | 'draft' | 'pending_verification' | 'approved' | 'rejected' | 'cancelled';

export type TradeInEstimate = {
  estimatedCredit: number;
  currency: string;
  estimatedPriceAfterTradeIn?: number | null;
  status: TradeInStatus;
};

export type TradeIn = TradeInEstimate & {
  id: number;
  category: string;
  brandModel: string;
  condition: TradeInCondition;
  imagePath: string;
  targetProductSlug?: string | null;
  targetProductName?: string | null;
  targetProductPrice?: number | null;
  handoffMethod: TradeInHandoffMethod;
  adminNote?: string | null;
  createdAt: string;
};

export type AppliedCoupon = {
  code: string;
  discountType: 'percentage' | 'fixed_amount';
  value: number;
  minimumSpend: number;
  discountAmount: number;
  badge: string;
};

export type Coupon = {
  id: number;
  name: string;
  code: string;
  discountType: 'percentage' | 'fixed_amount';
  value: number;
  minimumSpend: number;
  isActive: boolean;
  startsAt?: string | null;
  expiresAt?: string | null;
  usageLimit?: number | null;
  redemptionCount: number;
  createdAt: string;
};

export type OrderStatus = 'pending' | 'processing' | 'shipped' | 'delivered' | 'cancelled' | 'refunded';
export type PaymentStatus = 'unpaid' | 'paid' | 'failed' | 'refunded';

export type OrderItem = {
  id: number;
  productName: string;
  productSlug: string;
  sku: string;
  color: string;
  image?: string | null;
  quantity: number;
  unitPrice: number;
  lineTotal: number;
  giftBoxKey?: string | null;
  giftMessage?: string | null;
  packagingNotes?: string | null;
  surpriseRecipient?: string | null;
  surpriseVibes?: string[];
  surpriseInstructions?: string | null;
};

export type WalletTransaction = {
  id: number;
  type: 'exchange_credit' | 'checkout_debit' | 'order_reversal' | 'admin_adjustment';
  amount: number;
  balanceAfter: number;
  description: string;
  referenceType: string;
  referenceId: string;
  createdAt: string;
};

export type Wallet = {
  balance: number;
  currency: string;
  transactions: WalletTransaction[];
};

export type ExchangeStatus = 'pending_verification' | 'approved' | 'rejected' | 'cancelled' | 'completed';

export type ExchangeRequest = {
  id: number;
  orderItemId: number;
  originalProductName: string;
  newProductVariantId: number;
  newProductName: string;
  newProductSlug: string;
  newProductColor: string;
  oldProductValue: number;
  newProductValue: number;
  difference: number;
  walletCredit: number;
  amountDue: number;
  currency: string;
  invoiceIntact: boolean;
  packagingIntact: boolean;
  customerNote?: string | null;
  adminNote?: string | null;
  status: ExchangeStatus;
  createdAt: string;
  reviewedAt?: string | null;
  settlementOrderNumber?: string | null;
};

export type Order = {
  id: number;
  number: string;
  email: string;
  customerName: string;
  status: OrderStatus;
  paymentStatus: PaymentStatus;
  currency: string;
  subtotal: number;
  discountTotal: number;
  couponDiscountTotal: number;
  tradeInCredit: number;
  walletCredit: number;
  shippingTotal: number;
  total: number;
  couponCode?: string | null;
  couponDiscountType?: string | null;
  tradeIn?: TradeIn | null;
  channel: string;
  items: OrderItem[];
  itemCount: number;
  createdAt: string;
  paidAt?: string | null;
  deliveredAt?: string | null;
  exchangeEligibleUntil?: string | null;
  returnEligibleUntil?: string | null;
};

export type SiteSettings = {
  siteName: string;
  logoPath?: string | null;
  heroEyebrow: string;
  heroEyebrowTr?: string | null;
  heroHeadline: string;
  heroHeadlineTr?: string | null;
  heroSubheadline: string;
  heroSubheadlineTr?: string | null;
  heroImagePath?: string | null;
  contactEmail: string;
  contactPhone: string;
  contactAddress: string;
  instagramUrl?: string | null;
  tiktokUrl?: string | null;
  whatsappPhone?: string | null;
  pinterestUrl?: string | null;
  aboutHeadline: string;
  aboutHeadlineTr?: string | null;
  aboutBody: string;
  aboutBodyTr?: string | null;
};

export type AuthUser = {
  id: string;
  email: string;
  firstName: string;
  lastName: string;
  dateJoined: string;
  role: 'customer' | 'staff' | 'admin';
};

export type AuthSession = {
  access: string;
  refresh: string;
  user: AuthUser;
  accessExpiresAt: string;
};

export type AdminUser = AuthUser & {
  orderCount: number;
  lifetimeSpend: number;
  lastActiveAt?: string | null;
};

export type RevenuePoint = { name: string; date: string; revenue: number; orders: number };

export type TopProduct = {
  id: number;
  name: string;
  slug: string;
  image?: string | null;
  unitsSold: number;
  revenue: number;
  price: number;
};

export type Dashboard = {
  netRevenue: number;
  revenueChangePct: number;
  orderCount: number;
  orderChangePct: number;
  averageOrderValue: number;
  aovChangePct: number;
  customerCount: number;
  returningCustomerPct: number;
  activeProducts: number;
  lowStockCount: number;
  outOfStockCount: number;
  inventoryValue: number;
  revenueSeries: RevenuePoint[];
  recentOrders: Order[];
  topProducts: TopProduct[];
};
export type PopupType = 'promotional' | 'newsletter' | 'announcement' | 'support_care' | 'custom';
export type PopupPlacement = 'center_modal' | 'bottom_bar' | 'slide_in_bottom_right' | 'slide_in_bottom_left';
export type PopupTriggerType = 'delay' | 'scroll_depth' | 'exit_intent' | 'immediate';
export type PopupTargetAudience = 'all' | 'guests_only' | 'registered_only';

export type Popup = {
  id: number;
  type: PopupType;
  placement: PopupPlacement;
  triggerType: PopupTriggerType;
  triggerValue: number;
  cooldownDays: number;
  priority: number;
  targetPages?: string | null;
  deviceTarget: 'all' | 'desktop' | 'mobile';
  badge?: string | null;
  badgeTr?: string | null;
  title: string;
  titleTr?: string | null;
  description?: string | null;
  descriptionTr?: string | null;
  imagePath?: string | null;
  primaryCtaText?: string | null;
  primaryCtaTextTr?: string | null;
  primaryCtaUrl?: string | null;
  secondaryCtaText?: string | null;
  secondaryCtaTextTr?: string | null;
  couponCode?: string | null;
};

export type StorefrontPopup = Popup;

export type AdminPopup = Popup & {
  name: string;
  targetAudience: PopupTargetAudience;
  isActive: boolean;
  startsAt?: string | null;
  expiresAt?: string | null;
  impressionCount: number;
  clickCount: number;
  conversionCount: number;
  clickThroughRate: number;
  conversionRate: number;
  createdAt: string;
  updatedAt: string;
};

export type PopupDraft = {
  name: string;
  type: PopupType;
  placement: PopupPlacement;
  triggerType: PopupTriggerType;
  triggerValue: number;
  targetAudience: PopupTargetAudience;
  targetPages?: string | null;
  deviceTarget: 'all' | 'desktop' | 'mobile';
  cooldownDays: number;
  priority: number;
  isActive: boolean;
  startsAt?: string | null;
  expiresAt?: string | null;
  badge?: string | null;
  badgeTr?: string | null;
  title: string;
  titleTr?: string | null;
  description?: string | null;
  descriptionTr?: string | null;
  primaryCtaText?: string | null;
  primaryCtaTextTr?: string | null;
  primaryCtaUrl?: string | null;
  secondaryCtaText?: string | null;
  secondaryCtaTextTr?: string | null;
  couponCode?: string | null;
};

export type ChannelAttribution = { channel: string; orders: number; revenue: number; sharePct: number };
export type FunnelStep = { step: string; count: number };
export type CohortRow = { cohortWeek: string; cohortSize: number; retentionPcts: number[] };

export type Marketing = {
  totalSessions: number;
  sessionsChangePct: number;
  conversionRate: number;
  conversionChangePct: number;
  revenue: number;
  revenueChangePct: number;
  acquisitionCost: number;
  roasMultiplier: number;
  attribution: ChannelAttribution[];
  funnel: FunnelStep[];
  cohorts: CohortRow[];
  revenueSeries: RevenuePoint[];
};

File: api/src/MersTassel.Domain/Entities/Commerce.cs
using MersTassel.Domain.Common;
using MersTassel.Domain.Enums;

namespace MersTassel.Domain.Entities;

public class Cart : SoftDeletableEntity
{
    public string UserId { get; set; } = string.Empty;
    public AppUser User { get; set; } = null!;

    public string Email { get; set; } = string.Empty;
    public CartStatus Status { get; set; } = CartStatus.Open;
    public string Currency { get; set; } = "USD";

    public int? CouponId { get; set; }
    public Coupon? Coupon { get; set; }

    public ICollection<CartItem> Items { get; set; } = new List<CartItem>();
    public TradeInRequest? TradeIn { get; set; }
}

public class Coupon : SoftDeletableEntity
{
    public string Name { get; set; } = string.Empty;
    public string Code { get; set; } = string.Empty;
    public string NormalizedCode { get; set; } = string.Empty;
    public CouponDiscountType DiscountType { get; set; }
    public decimal Value { get; set; }
    public decimal MinimumSpend { get; set; }
    public bool IsActive { get; set; } = true;
    public DateTimeOffset? StartsAt { get; set; }
    public DateTimeOffset? ExpiresAt { get; set; }
    public int? UsageLimit { get; set; }
    public int RedemptionCount { get; set; }
}

public class CartItem : SoftDeletableEntity
{
    public int CartId { get; set; }
    public Cart Cart { get; set; } = null!;

    public int ProductVariantId { get; set; }
    public ProductVariant Variant { get; set; } = null!;

    public int Quantity { get; set; } = 1;

    /// <summary>
    /// Items created by the Kavanoz builder share a key so they remain a coherent gift box
    /// even when the shopper has ordinary products in the same cart.
    /// </summary>
    public string? GiftBoxKey { get; set; }
    public string? GiftMessage { get; set; }
    public string? PackagingNotes { get; set; }
}

public class Order : SoftDeletableEntity
{
    public string Number { get; set; } = string.Empty;

    /// <summary>Null once a customer account is removed; the order itself is retained.</summary>
    public string? UserId { get; set; }
    public AppUser? User { get; set; }

    public string Email { get; set; } = string.Empty;
    public string CustomerName { get; set; } = string.Empty;

    public OrderStatus Status { get; set; } = OrderStatus.Pending;
    public PaymentStatus PaymentStatus { get; set; } = PaymentStatus.Unpaid;

    public string Currency { get; set; } = "USD";
    public decimal Subtotal { get; set; }
    public decimal DiscountTotal { get; set; }
    public decimal TradeInCredit { get; set; }
    public decimal WalletCredit { get; set; }
    public decimal ShippingTotal { get; set; }
    public decimal Total { get; set; }
    public string? CouponCode { get; set; }
    public string? CouponDiscountType { get; set; }

    /// <summary>Serialized shipping address; SQLite has no native JSON column type.</summary>
    public string ShippingAddressJson { get; set; } = "{}";

    public string Channel { get; set; } = "storefront";
    public string? IdempotencyKey { get; set; }

    public string? StripeCheckoutSessionId { get; set; }
    public string? StripePaymentIntentId { get; set; }
    public DateTimeOffset? PaidAt { get; set; }
    public DateTimeOffset? DeliveredAt { get; set; }

    public ICollection<OrderItem> Items { get; set; } = new List<OrderItem>();
    public ICollection<InventoryReservation> Reservations { get; set; } = new List<InventoryReservation>();
    public TradeInRequest? TradeIn { get; set; }
}

/// <summary>A currency-specific customer store-credit account backed by an immutable ledger.</summary>
public class StoreWallet
{
    public int Id { get; set; }
    public string UserId { get; set; } = string.Empty;
    public AppUser User { get; set; } = null!;
    public string Currency { get; set; } = "USD";
    public decimal Balance { get; set; }
    public string ConcurrencyStamp { get; set; } = Guid.NewGuid().ToString("N");
    public DateTimeOffset CreatedAt { get; set; } = DateTimeOffset.UtcNow;
    public DateTimeOffset UpdatedAt { get; set; } = DateTimeOffset.UtcNow;
    public ICollection<StoreWalletTransaction> Transactions { get; set; } = new List<StoreWalletTransaction>();
}

/// <summary>Append-only wallet movement; Amount is positive for credit and negative for debit.</summary>
public class StoreWalletTransaction
{
    public int Id { get; set; }
    public int WalletId { get; set; }
    public StoreWallet Wallet { get; set; } = null!;
    public WalletTransactionType Type { get; set; }
    public decimal Amount { get; set; }
    public decimal BalanceAfter { get; set; }
    public string Description { get; set; } = string.Empty;
    public string ReferenceType { get; set; } = string.Empty;
    public string ReferenceId { get; set; } = string.Empty;
    public string IdempotencyKey { get; set; } = string.Empty;
    public DateTimeOffset CreatedAt { get; set; } = DateTimeOffset.UtcNow;
}

/// <summary>
/// A request to exchange one unit from a delivered order for a current catalog variant.
/// Values are snapshotted server-side so later price edits cannot change the agreement.
/// </summary>
public class ExchangeRequest : SoftDeletableEntity
{
    public string UserId { get; set; } = string.Empty;
    public AppUser User { get; set; } = null!;
    public int OrderItemId { get; set; }
    public OrderItem OrderItem { get; set; } = null!;
    public int NewProductVariantId { get; set; }
    public ProductVariant NewProductVariant { get; set; } = null!;
    public decimal OldProductValue { get; set; }
    public decimal NewProductValue { get; set; }
    public decimal Difference { get; set; }
    public decimal WalletCredit { get; set; }
    public decimal AmountDue { get; set; }
    public string Currency { get; set; } = "USD";
    public bool InvoiceIntact { get; set; }
    public bool PackagingIntact { get; set; }
    public string? CustomerNote { get; set; }
    public string? AdminNote { get; set; }
    public ExchangeRequestStatus Status { get; set; } = ExchangeRequestStatus.PendingVerification;
    public DateTimeOffset? ReviewedAt { get; set; }
    public int? WalletTransactionId { get; set; }
    public StoreWalletTransaction? WalletTransaction { get; set; }
    public int? SettlementOrderId { get; set; }
    public Order? SettlementOrder { get; set; }
}

/// <summary>
/// A provisional credit request. The estimate reduces the checkout amount immediately, but
/// remains pending until the atelier physically inspects the submitted item.
/// </summary>
public class TradeInRequest : SoftDeletableEntity
{
    public string? UserId { get; set; }
    public AppUser? User { get; set; }

    public int? CartId { get; set; }
    public Cart? Cart { get; set; }

    public int? OrderId { get; set; }
    public Order? Order { get; set; }

    public string Category { get; set; } = string.Empty;
    public string BrandModel { get; set; } = string.Empty;
    public TradeInCondition Condition { get; set; }
    public string ImagePath { get; set; } = string.Empty;
    public string? TargetProductSlug { get; set; }
    public string? TargetProductName { get; set; }
    public decimal? TargetProductPrice { get; set; }
    public decimal EstimatedCredit { get; set; }
    public string Currency { get; set; } = "USD";
    public TradeInHandoffMethod HandoffMethod { get; set; }
    public TradeInStatus Status { get; set; } = TradeInStatus.PendingVerification;
    public string? AdminNote { get; set; }
}

/// <summary>
/// An immutable snapshot of what was bought. Name, SKU and price are copied at checkout so
/// later catalog edits never rewrite the historical record.
/// </summary>
public class OrderItem : SoftDeletableEntity
{
    public int OrderId { get; set; }
    public Order Order { get; set; } = null!;

    public int? ProductVariantId { get; set; }
    public ProductVariant? Variant { get; set; }

    public string ProductName { get; set; } = string.Empty;
    public string ProductSlug { get; set; } = string.Empty;
    public string Sku { get; set; } = string.Empty;
    public string Color { get; set; } = string.Empty;
    public string? ImagePath { get; set; }
    public int Quantity { get; set; }
    public decimal UnitPrice { get; set; }

    /// <summary>Immutable Kavanoz instructions copied from the cart at checkout.</summary>
    public string? GiftBoxKey { get; set; }
    public string? GiftMessage { get; set; }
    public string? PackagingNotes { get; set; }
}

public class InventoryReservation : SoftDeletableEntity
{
    public int OrderId { get; set; }
    public Order Order { get; set; } = null!;

    public int ProductVariantId { get; set; }
    public ProductVariant Variant { get; set; } = null!;

    public int Quantity { get; set; }
    public ReservationStatus Status { get; set; } = ReservationStatus.Active;
    public DateTimeOffset ExpiresAt { get; set; }
}

/// <summary>Webhook idempotency ledger — Stripe redelivers events, so each id is processed once.</summary>
public class ProcessedStripeEvent
{
    public int Id { get; set; }
    public string EventId { get; set; } = string.Empty;
    public string EventType { get; set; } = string.Empty;
    public DateTimeOffset ProcessedAt { get; set; } = DateTimeOffset.UtcNow;
}
